### ISSUE TITLE
Factoring Out all Project Specific Values

### ISSUE THREAD

**Issue by softworkz** [gitgitgadget/gitgitgadget#829]
At this point, I have just plainly replaced all ggg specific parameters in my fork, which is undoubtedly not a great way for moving forward.

As I already know all the places that require parametrization, I would get this done and contribute that change. I'd just like to avoid doing it more than once, so it would be great when we could agree about a pattern for this up-front.
I don't have any preferences for this, other than it should probably be a single file at the end where you can do the adaption for using ggg in another project.

Maybe you can just make a suggestion?

**Comment by dscho** [comment]
> Maybe you can just make a suggestion?

So far, we did a pretty good job of allowing things to be configured via the `[gitgitgadget]` section of the Git config...

**Comment by softworkz** [comment]
Those few things yes - but I'm talking about the many things that are hardcoded now but need to be configurable.
Look at project-options.ts. I think there's a reason why you didn't make all this configurable via git config..?

**Comment by dscho** [comment]
> there's a reason why you didn't make all this configurable via git config..?

Yes. The reason is that all of GitGitGadget used to be [a shell script](https://github.com/dscho/mail-patch-series), with no configuration whatsoever. We made configurable what we needed to be configurable. And now we need more things to be configurable.

**Comment by softworkz** [comment]
So you want to have it all in the Git config? OK we can do that.

**Comment by webstech** [comment]
I'm not sure the git config is the best place.  A json file may be better.  Future changes requiring git config changes would be a problem for random contributors.

This is project level configuration that users should not have to manage.  The current ggg related git config values tend to be user specific.

**Comment by softworkz** [comment]
From my point of view: the more direct and connected to the code - the better.
Perfect would be something that that connects even at dev-time - i.e. some form of code file (basically in a way like this: https://github.com/gitgitgadget/gitgitgadget/issues/818#issuecomment-1006054970).

The advantage of this would be to have the ability to see how things are connected and which values are configured already during development, just by navigating through the source code (when the file would be included in your IDE).

One problem with this project is that the actual execution is happening somewhere "far away" which makes it impossible to debug and read runtime values. 
When having an indirect mechanism like via a config file, and there's some issue to troubleshoot, you often want to verify that the right values are loaded from configuration. This is somewhat tedious because it forces you to add some code to dump/log those values to verify. Besides that, you need to be careful which values to dump, because the logs are publicly visible...

For those reasons, a configuration that is part of the code would be my preference. 
Or better - let's not call it "configuration" at all. This is more about "code/project customizations", which might possibly even include an alternate implementation of a function (e.g. like in project-options.ts).

But I'm open to all other ideas as well.

**Comment by dscho** [comment]
> I'm not sure the git config is the best place. A json file may be better.

The big advantage of the Git config is that you can easily configure it via the environment variable `GIT_CONFIG_PARAMETERS`. That's definitely a nice way to configure it in GitHub workflows/Azure Pipelines.

But yes, the canonical way to customize a project like GitGitGadget would be via a JSON file.

> Perfect would be something that connects even at dev-time - i.e. some form of code file (basically in a way like this: [#818 (comment)](https://github.com/gitgitgadget/gitgitgadget/issues/818#issuecomment-1006054970)).

That would work, but would require _all_ projects using GitGitGadget to fork _and maintain their fork_.

**Comment by softworkz** [comment]
> That would work, but would require _all_ projects using GitGitGadget to fork _and maintain their fork_.

Even after having parametrized all the things where "we" need customization, I strongly doubt that this will get any time soon to a point where it can be used as an "out-of-the-box" solution. 
I doubt that it will make sense for any adopter to use this _without_ maintaining an own fork.

Just thinking about all the things that I've gone through to get it all set up and working and adapt to the requirements and the way things are set up. 
Just consider a case where you make a change that an adopter doesn't want. I don't think anybody would want to take the risk of auto-updating from your repo on one side - and on the other side, being unable to customize something that isn't customizable.

Or think about auto-updating: the way I've set this up allows me to update the whole system just by pushing an update to the main branch. For having that ability alone, I would always want to set up and maintain my own fork.

But the primary point is: for setting this up, you need to deep-dive into many things - including the code.
We can make this a lot more friendly for adoption through others by:

- improving documentation (most  importantly pipelines setup, but also many other parts)
- factoring out all the hard-coded ggg-specific stuff from the source code to ease the adoption procedure by a magnitude

but in the end this won't suffice that it could be taken "as-is". I don't see it as a "shelf-product" where you just need to do some configuration to get going, but more like a toolkit that requires a certain amount of work to adopt (but eventually a lot less than what I have gone through).

**Comment by dscho** [comment]
> I strongly doubt that this will get any time soon to a point where it can be used as an "out-of-the-box" solution.

Me, too. Nevertheless, it would be nice if we did not throw the door shut on such an end game.

**Comment by webstech** [comment]
Decided to play with this a bit using @softworkz changes.  It is just a POC at https://github.com/webstech/gitgitgadget/tree/generify.

There is a commit with updates to the shell scripts to support environment variables to control more settings.  Of course the scripts have some very GGG checks.  Not sure how much (if any) use they get anymore.  Untested and maybe more than a few errors at this point.

The second commit is an initial POC to use a config file to make GGG generic.  This is an alternative to trying to package GGG.  The alternative config (and workflows?) would be in the target project (eg git/git) or a separate project (more likely).  Any workflow would clone the separate project and GGG.  The changes somewhat cover `misc-helper.ts` and `ci-helper.ts` based on the blanket changes. So far, unit tests are still working.

There are some very `git` specific checks in the code so a mechanism would be needed to abstract them out.

Is this a direction worth pursuing?

**Comment by dscho** [comment]
> a POC at https://github.com/webstech/gitgitgadget/tree/generify.

That looks like a good first draft.

Before we pursue this further, I think we should probably lift the line length restriction of GitGitGadget's own source code (or at least extend it to 120 columns/line).

**Comment by dscho** [comment]
> Is this a direction worth pursuing?

Oh, forgot to answer this question. I am not quite sure. An idea I had recently was that we could have a class holding that configuration, but initializing it based on hard-coded values depending on the repository we're targeting.

**Comment by softworkz** [comment]
@webstech - Excellent! I like it at lot. It's pretty much the same that I had in mind 😄 
@dscho - +1 for lifting the column limit (or increasing). The problem is that with all the structural indentation, there's often not much usable space left. 

> An idea I had recently was that we could have a class holding that configuration, but initializing it based on hard-coded values depending on the repository we're targeting.

Isn't that possible in the same way with @webstech's proposal? An "installation" of GGG can already target multiple repositories (when they are related/forked).

On the other side, multiple unrelated repositories cannot reasonably be served by a single "installation" of GGG anyway, because it would also require different pipelines and repo paths. What I mean is: the biggest effort is required for setting up the pipelines. Having a one-serves-many "installation" of GGG would not make the setup process much easier (maybe even more complicated, though).
